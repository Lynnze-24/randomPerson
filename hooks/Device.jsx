import { Platform} from "react-native";

export function useDevice() {
  const { OS: platform } = Platform;

 

  const isWeb = platform === "web";
  const isAndroid = platform === "android";
  const isIOS = platform === "ios";

  return {
    isWeb,
    isAndroid,
    isIOS,
    
  };
}