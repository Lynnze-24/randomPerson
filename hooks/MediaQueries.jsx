import React,{ useMemo } from 'react';
import { useWindowDimensions } from 'react-native'

const useMediaQueries = () => {
    const {  width } = useWindowDimensions();
   
    const isLessThanXS = useMemo(() => {
        return width <= 465;
      }, [width]);
    
      const isLessThanSM = useMemo(() => {
        return width <= 550;
      }, [width]);

      const isLessThanXSM = useMemo(() => {
        return width <= 600;
      }, [width]);
    
      const isLessThanMD = useMemo(() => {
        return width <= 700;
      }, [width]);
    
    
      const isLessThanLG = useMemo(() => {
        return width <= 840;
      }, [width]);

      const isLessThanXLG = useMemo(() => {
        return width <= 1000;
      }, [width]);

      const isLessThanXL = useMemo(() => {
        return width <= 1260;
      }, [width]);

      
    

      return {
        width,
        isLessThanXS,
        isLessThanSM,
        isLessThanXSM,
        isLessThanMD,
        isLessThanLG,
        isLessThanXLG,
        isLessThanXL,
      };
}

export default useMediaQueries;
