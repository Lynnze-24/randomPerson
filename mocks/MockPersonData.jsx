export const mockPersonData = [
    {
        name:{
            title:"Mr",
            first:"Aung",
            last:"Linn"
        },
        picture:{
            thumbnail:1
        },
        email:'bbb@gmail.com',
        cell:"999",
        location:{
            city:"Bago",
            country:"Myanmar"
        },
        login:{
            md5:122
        }
        
    },
    {
        name:{
            title:"Mrs",
            first:"Htet",
            last:"Htet"
        },
        picture:{
            thumbnail:1
        },
        email:'aaa@gmail.com',
        cell:"999",
        location:{
            city:"Bago",
            country:"Myanmar"
        },
        login:{
            md5:333
        }
        
    }
]