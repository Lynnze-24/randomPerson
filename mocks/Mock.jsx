import { mockPersonData } from "./MockPersonData"

export const mockVal = {personData:[],
    isLoading:false,
    isError:false,
    refreshPerson:jest.fn(),
    filter:'',
    setFilter:jest.fn(),
    refresh:false,
    isTableView:true,
    switchView:jest.fn()
}




export const mockPersonVal = {...mockVal,
    personData: mockPersonData

}