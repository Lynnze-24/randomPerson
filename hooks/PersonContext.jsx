import React, { createContext, useContext, useState,useEffect} from 'react';
import { mockPersonData } from '../mocks/MockPersonData';



const PersonContext = createContext();


export function usePersonContext(){
    return useContext(PersonContext)
}

export function PersonProvider({children,mock,mockFetch}) {
    const initialData = mockFetch?mockPersonData:[]
    const initialLoading = mockFetch?false:true
   
   const [personData, setPersonData] = useState(initialData); // for person data
    
    const[isLoading,setIsLoading] = useState(initialLoading); // for loading state
    const [isError,setIsError] = useState(false);   // handle fetch error 
    const[refresh,setRefresh] = useState(false);    // for refresh button
    const[filter,setFilter] = useState('');         // setting filter word to filter person data
    
    const[isTableView,setTableView] = useState(true) // switch view

    const switchView = () =>{
        setTableView(!isTableView)
    }

    const fetchPerson = ((mock||mockFetch))?()=>{} 
                            :()=>{
                                fetch('https://randomuser.me/api?results=50')
                                .then(res=> res.json())
                                .then(person=> {
                                        if(isError)setIsError(false)
                                        setPersonData(person.results)
                                        setIsLoading(false)
                                    
                                })
                                .catch(()=>{
                                    setIsLoading(false)
                                    setIsError(true);
                                
                                })
                        }
  

    const refreshPerson = ()=>{
        setIsLoading(true);
        setRefresh(!refresh)
    }
      
    useEffect(() => {
           fetchPerson()
    }, [refresh])
    
    const val = mock? mock : {personData,
                                isLoading,
                                isError,
                                refreshPerson,
                                filter,
                                setFilter,
                                refresh,
                                isTableView,
                                switchView}
    
    return (
        <PersonContext.Provider value={val}>
        {children}
      </PersonContext.Provider>
    );
}
