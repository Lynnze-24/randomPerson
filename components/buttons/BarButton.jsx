import React,{useState} from 'react'
import { IconButton } from 'react-native-paper'
import { StyleSheet } from 'react-native'
import cusTheme from '../../theme/Theme'
import { usePersonContext } from '../../hooks/PersonContext'
import useMediaQueries from '../../hooks/MediaQueries'

const BarButton = ({icon,type,onPress,testID}) => {
    const{isLoading} = usePersonContext()
    const {isLessThanSM} = useMediaQueries()
    
    return (
        <IconButton
            testID={testID}
            style={[styles.container,isLessThanSM && styles.sm]}
            icon={icon}
            disabled={icon==='refresh'?isLoading:false}
            type={type}
            color={cusTheme.secondary}
            size={20}
            onPress={onPress}
        />
    )
}

const styles = StyleSheet.create({
    container: {
        marginRight:30,
        marginLeft:0
   },
   sm:{
    marginRight:10
   }
})

export default BarButton;
