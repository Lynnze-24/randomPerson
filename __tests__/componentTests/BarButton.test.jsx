import {render, fireEvent } from "@testing-library/react-native";
import BarButton from "../../components/buttons/BarButton";
import { PersonProvider } from "../../hooks/PersonContext";
import { mockVal } from "../../mocks/Mock";
import renderer from 'react-test-renderer';

const mockPress= jest.fn()


it('matches snapShot', () => {
    const comp = renderer.create( <PersonProvider mock={mockVal}>
                                <BarButton onPress={mockPress} testID="test"/>
                                </PersonProvider>).toJSON();
    expect(comp).toMatchSnapshot();
});



it('should render component correctly',()=>{ 
    const {getByTestId}=render(
                        <PersonProvider mock={mockVal}>
                        <BarButton onPress={mockPress} testID="test"/>
                        </PersonProvider>
                        );
    const barBtn =getByTestId("test")  // check if component renders
    fireEvent.press(barBtn)
    expect(mockPress).toHaveBeenCalled() // check onPress call
})

it('should disable according to loading state if refresh icon',()=>{ 
    const mock2 = {...mockVal,isLoading:true} // change loading state to true
    const {getByTestId,debug}=render(
                        <PersonProvider mock={mock2}>
                        <BarButton icon="refresh" onPress={mockPress} testID="test"/>
                        </PersonProvider>
                        );
        expect(getByTestId('test')).toHaveProp('accessibilityState',{"disabled": true})
     
})