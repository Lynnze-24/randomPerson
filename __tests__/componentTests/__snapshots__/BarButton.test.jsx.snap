// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapShot 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  hitSlop={
    Object {
      "bottom": 6,
      "left": 6,
      "right": 6,
      "top": 6,
    }
  }
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "margin": 6,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 15,
          "height": 30,
          "width": 30,
        },
        false,
        Array [
          Object {
            "marginLeft": 0,
            "marginRight": 30,
          },
          false,
        ],
      ],
    ]
  }
  testID="test"
>
  <View />
</View>
`;
